### DISCUSSION TITLE
UnicodeDecodeError when running inference from the command line

### DISCUSSION THREAD

**Discussion by Cong-SUN** [talmolab/sleap#1360] [Help!] [answered]
Heyy, I used this 
`(sleap) C:\Windows\system32>sleap-track "C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4" --frames 0-1494 -m "C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/models/test_11_230621_145849.single_instance.n=260/best_model.h5" [other_args] --peak_threshold 0.05` 
to run inference but it gives this UnicodeDecodeError:

```shell
Video: C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4
Traceback (most recent call last):
  File "C:\casa_phy\Miniconda3\envs\sleap\Scripts\sleap-track-script.py", line 33, in <module>
    sys.exit(load_entry_point('sleap==1.3.0', 'console_scripts', 'sleap-track')())
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\site-packages\sleap\nn\inference.py", line 5334, in main
    predictor = _make_predictor_from_cli(args)
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\site-packages\sleap\nn\inference.py", line 5231, in _make_predictor_from_cli
    progress_reporting=args.verbosity,
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\site-packages\sleap\nn\inference.py", line 4791, in load_model
    resize_input_layer=resize_input_layer,
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\site-packages\sleap\nn\inference.py", line 221, in from_model_paths
    model_configs = [sleap.load_config(model_path) for model_path in model_paths]
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\site-packages\sleap\nn\inference.py", line 221, in <listcomp>
    model_configs = [sleap.load_config(model_path) for model_path in model_paths]
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\site-packages\sleap\nn\config\training_job.py", line 158, in load_config
    filename, load_training_config=load_training_config
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\site-packages\sleap\nn\config\training_job.py", line 120, in load_json
    json_data = f.read()
  File "C:\casa_phy\Miniconda3\envs\sleap\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 40: character maps to <undefined>```

**Comment by Cong-SUN** [comment]
According to ChatGPT:

The UnicodeDecodeError you encountered suggests that the file you are trying to read contains characters that cannot be decoded using the 'cp1252' encoding, which is the default encoding for the 'charmap' codec on Windows.

To resolve this issue, you can try the following steps:

Use a different encoding: Instead of using the default 'cp1252' encoding, try using a different encoding that supports a broader range of characters, such as 'utf-8' or 'latin-1'. You can specify the encoding explicitly when reading the file.

Specify the encoding parameter explicitly: When opening the file, use the encoding parameter and set it to the appropriate encoding for your file. 

---- but I don't know how to specify the encoding parameters?

**Comment by roomrys** [comment]
Hi @Cong-SUN,

I am just double checking that none of the files referenced (either video or model) are contained in a zipped directory?

Thanks,
Liezl

**Reply by Cong-SUN** [reply]
Yep no, they are not zipped. Thank you!

**Comment by roomrys** [comment, marked as answer]
Hi @Cong-SUN,

I think I see a potential problem. For the model path, can you feed in just the path to the model directory (SLEAP looks for the training_config.json within this directory). Afterwards, can you post the new command you tried and let me know how it goes?

Thanks,
Liezl

**Reply by Cong-SUN** [reply]
Yes I think that might be the problem! I tried `sleap-track "C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4" --frames 0-1493 -m "C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/models/test_11_230621_145849.single_instance.n=260" [other_args] --peak_threshold 0.05` and it worked:
```shell
Finished inference at: 2023-06-23 19:17:32.379806
Total runtime: 33.899048805236816 secs
Predicted frames: 1494/1494
Provenance:
{
│   'model_paths': ['C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/models/test_11_230621_145849.single_instance.n=260\\training_config.json'],
│   'predictor': 'SingleInstancePredictor',
│   'sleap_version': '1.3.0',
│   'platform': 'Windows-10-10.0.17763-SP0',
│   'command': 'C:\\casa_phy\\Miniconda3\\envs\\sleap\\Scripts\\sleap-track C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4 --frames 0-1493 -m C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/models/test_11_230621_145849.single_instance.n=260 [other_args] --peak_threshold 0.05',
│   'data_path': 'C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4',
│   'output_path': 'C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4.predictions.slp',
│   'total_elapsed': 33.899048805236816,
│   'start_timestamp': '2023-06-23 19:16:58.480758',
│   'finish_timestamp': '2023-06-23 19:17:32.379806'
}

Saved output: C:/Users/Branco Lab/.sleap/1.3.0/Yaara_patch_analysis_csun/videos_mp4/230502_CA1103258_Patch(0)-FireWire-2.mp4.predictions.slp
```
Thank you so much!
